daemon/plugin_p.cpp

```diff
@@ -6,43 +6,105 @@
+    }
+
+    return autotestMode;
 }
 
-Daemon::ApiImpl::PluginHelper::~PluginHelper()
+Daemon::ApiImpl::PluginManager::PluginManager()
+    : m_autotestMode(isAutotestMode())
 {
 }
 
-Daemon::ApiImpl::PluginHelper::FailureType
-Daemon::ApiImpl::PluginHelper::failureType() const
+Daemon::ApiImpl::PluginManager *Daemon::ApiImpl::PluginManager::instance()
 {
-    return m_failureType;
+    if (!pluginManagerInstance) {
+        pluginManagerInstance = new PluginManager();
+    }
+
+    return pluginManagerInstance;
 }
 
-void Daemon::ApiImpl::PluginHelper::reportFailure(
-        const QLoggingCategory &category())
+QVector<QPluginLoader*> Daemon::ApiImpl::PluginManager::loadPluginFiles()
 {
-    if (m_failureType != PluginLoadFailure
-            && m_failureType != PluginTypeFailure) {
-        // already reported this error
-        return;
+    QVector<QPluginLoader*> result;
+    QStringList paths = QCoreApplication::libraryPaths();
+
+    Q_FOREACH(const QString &path, paths) {
+        // Don't enumerate /usr/bin
+        if (path == "/usr/bin")
+            continue;
+
+        QDir dir(path);
+        Q_FOREACH (const QFileInfo &file, dir.entryInfoList(QDir::Files | QDir::NoDot | QDir::NoDotDot, QDir::Name)) {
+            const QString fileName = file.fileName();
+
+            // Don't even try to load files which don't look like libraries
+            if (!fileName.startsWith("lib") || !fileName.contains(".so")) {
+                continue;
+            }
+
+            // load the plugin
+            auto *loader = new QPluginLoader(file.absoluteFilePath());
+            if (!loader->load()) {
+                qCWarning(lcSailfishSecretsPlugins) << "Could not load plugin:" << loader->fileName();
+                delete loader;
+                continue;
+            }
+
+            result.append(loader);
+        }
+    }
+
+    return result;
+}
+
+void Daemon::ApiImpl::PluginManager::addPlugin(QPluginLoader *loader, const PluginHelpers::PluginInfo &info, QObject *obj)
+{
+    bool use = true;
+
+    if (!info.canUse) {
+        qCWarning(lcSailfishSecretsPlugins) << "Not a usable crypto or secrets plugin:" << loader->fileName();
+        use = false;
+    }
+
+    if (m_plugins.contains(info.name)) {
+        qCWarning(lcSailfishSecretsPlugins) << "Not adding plugin with duplicate name:" << loader->fileName();
+        use = false;
+    }
+
+    if (info.name.endsWith(QStringLiteral(".test"), Qt::CaseInsensitive) != m_autotestMode) {
+        qCWarning(lcSailfishSecretsPlugins) << "Not adding plugin because of testing mode mismatch:" << loader->fileName();
+        use = false;
     }
 
-    if (!isLoaded()) {
-        qCWarning(category) << "cannot load plugin:"
-                            << fileName() << errorString();
+    if (!use) {
+        if (!loader->unload()) {
+            qCWarning(lcSailfishSecretsPlugins) << "Could not unload plugin:" << loader->fileName();
+        }
+        delete loader;
         return;
     }
```

Comment:
Maybe add an else block with: { qCDebug() << "Adding plugin:" << loader->fileName(); delete loader; }

Reply:
Or perhaps "Adding plugin:" << loader->fileName() << "as" << info.name; for maximum information.